pan3d/dataset_viewer.py

```diff
@@ -0,0 +1,485 @@
+import asyncio
+import concurrent.futures
+import json
+import pandas
+import pyvista
+from pathlib import Path
+from typing import Optional, Union
+
+from trame.decorators import TrameApp, change
+from trame.app import get_server
+from trame.widgets import html, client
+from trame.widgets import vuetify3 as vuetify
+from trame_server.core import Server
+from trame_server.state import State
+from trame_server.controller import Controller
+from trame_vuetify.ui.vuetify3 import VAppLayout
+
+from pan3d.dataset_builder import DatasetBuilder
+from pan3d.ui import AxisDrawer, MainDrawer, Toolbar, RenderOptions
+from pan3d.utils import (
+    initial_state,
+    has_gpu_rendering,
+)
+
+BASE_DIR = Path(__file__).parent
+CSS_FILE = BASE_DIR / "ui" / "custom.css"
+
+
+@TrameApp()
+class DatasetViewer:
+    """Create a Trame GUI for a DatasetBuilder instance and manage rendering"""
+
+    def __init__(
+        self,
+        builder: Optional[DatasetBuilder] = None,
+        server: Union[Server, str] = None,
+        state: dict = None,
+        pangeo: bool = False,
+    ) -> None:
+        """Create an instance of the DatasetViewer class.
+
+        Parameters:
+            builder: Pan3D DatasetBuilder instance.
+            server: Trame server instance.
+            state:  A dictionary of initial state values.
+            pangeo: If true, use a list of example datasets from Pangeo Forge (examples/pangeo_catalog.json).
+        """
+        if builder is None:
+            builder = DatasetBuilder()
+            builder._viewer = self
+        self.builder = builder
+        self.server = get_server(server, client_type="vue3")
+        self.current_event_loop = asyncio.get_event_loop()
+        self.pool = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+        self._layout = None
+
+        self.plotter = pyvista.Plotter(off_screen=True, notebook=False)
+        self.plotter.set_background("lightgrey")
+        self.plot_view = None
+        self.actor = None
+        self.ctrl.get_plotter = lambda: self.plotter
+
+        self.state.update(initial_state)
+        self.state.ready()
+        if state:
+            self.state.update(state)
+        if pangeo:
+            with open(Path(BASE_DIR, "../examples/pangeo_catalog.json")) as f:
+                self.state.available_datasets += json.load(f)
+
+        self._force_local_rendering = not has_gpu_rendering()
+        if self._force_local_rendering:
+            pyvista.global_theme.trame.default_mode = "client"
+
+    def __repr_html__(self):
+        return self.layout
+
+    @property
+    def state(self) -> State:
+        """Returns the current State of the Trame server."""
+        return self.server.state
+
+    @property
+    def ctrl(self) -> Controller:
+        """Returns the Controller for the Trame server."""
+        return self.server.controller
+
+    @property
+    def layout(self) -> VAppLayout:
```

Comment:
So far on trame app that have a ui, we tend to call that property that return the layout just `ui`.
See https://kitware.github.io/trame/examples/core/jupyter.html

Reply:
Ok, changed in [b87a945](https://github.com/Kitware/pan3d/pull/59/commits/b87a9450a989c0c672d41afe7d852b030323d219)